src/Constraint/Constraint.php

```diff
@@ -128,13 +151,18 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
     }
 
     /**
-     * @param Constraint $provider
+     * @param ConstraintInterface $provider
      * @param bool $compareBranches
      *
      * @return bool
      */
-    public function matchSpecific(Constraint $provider, $compareBranches = false)
```

Comment:
Well the only reason it is there is Composer. It's used by `DefaultPolicy::versionCompare`. We could move the method to `Composer\Package\LinkConstraint\VersionConstraint`.